`AssertionError` with `payload Error Error.type UnresolvedSymbol<==>` when executing Base_Tests

More info has been added to the reported missing polyglot context assertion (#11113) which resulted in
```
 INFO ide_ci::program::command: cmd ⚠️ Caused by: java.lang.AssertionError: Failed to log failed conversion of null to string and payload Error Error.type UnresolvedSymbol<==>
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.computeMessage(PanicException.java:95)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.getMessage(PanicException.java:76)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.toString(PanicException.java:106)
 INFO ide_ci::program::command: cmd ⚠️ 	at java.base/java.lang.Throwable.<init>(Throwable.java:319)
 INFO ide_ci::program::command: cmd ⚠️ 	at java.base/java.lang.Exception.<init>(Exception.java:103)
 INFO ide_ci::program::command: cmd ⚠️ 	at java.base/java.io.IOException.<init>(IOException.java:98)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.IsolatedClassLoader//org.enso.runner.Main.withProfiling(Main.java:1205)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1450)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.IsolatedClassLoader//org.enso.runner.Main.launch(Main.java:1402)
 INFO ide_ci::program::command: cmd ⚠️ 	at org.enso.IsolatedClassLoader//org.enso.runner.Main.main(Main.java:1019)
 INFO ide_ci::program::command: cmd ⚠️ 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
 INFO ide_ci::program::command: cmd ⚠️ 	... 2 more
 INFO ide_ci::program::command: cmd ⚠️ Caused by: java.lang.AssertionError: No polyglot context is entered. A language or context reference must not be used if there is no polyglot context entered.
 INFO ide_ci::program::command: cmd ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotFastThreadLocals.assertValidGet(PolyglotFastThreadLocals.java:481)
 ```
- at https://github.com/enso-org/enso/actions/runs/10926101745/job/30329191172#step:7:1251
- at https://github.com/enso-org/enso/actions/runs/11036714355/job/30655819392?pr=10823#step:7:2605

Not sure if that's useful but I encountered such error on CI:
```
INFO ide_ci::program::command: bash ⚠️ Exception in thread "main" java.lang.AssertionError: Failed to log failed conversion of null to string and payload Error Error.type UnresolvedSymbol<==>
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.computeMessage(PanicException.java:95)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.getMessage(PanicException.java:76)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.toString(PanicException.java:106)
 INFO ide_ci::program::command: bash ⚠️ 	at java.base/java.lang.Throwable.<init>(Throwable.java:319)
 INFO ide_ci::program::command: bash ⚠️ 	at java.base/java.lang.Exception.<init>(Exception.java:103)
 INFO ide_ci::program::command: bash ⚠️ 	at java.base/java.io.IOException.<init>(IOException.java:98)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runner/org.enso.runner.Main.withProfiling(Main.java:1221)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runner/org.enso.runner.Main.launch(Main.java:1466)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runner/org.enso.runner.Main.launch(Main.java:1418)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runner/org.enso.runner.Main.main(Main.java:1033)
 INFO ide_ci::program::command: bash ⚠️ Caused by: java.lang.AssertionError: No polyglot context is entered. A language or context reference must not be used if there is no polyglot context entered.
 INFO ide_ci::program::command: bash ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotFastThreadLocals.assertValidGet(PolyglotFastThreadLocals.java:481)
 INFO ide_ci::program::command: bash ⚠️ 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotFastThreadLocals$ContextReferenceImpl.get(PolyglotFastThreadLocals.java:513)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.EnsoContext.get(EnsoContext.java:248)
 INFO ide_ci::program::command: bash ⚠️ 	at org.enso.runtime/org.enso.interpreter.runtime.error.PanicException.computeMessage(PanicException.java:91)
 INFO ide_ci::program::command: bash ⚠️ 	... 9 more
```

--- a/engine/runner/src/main/java/org/enso/runner/Main.java
+++ b/engine/runner/src/main/java/org/enso/runner/Main.java
@@ -742,23 +742,30 @@ public class Main {
       factory.checkForWarnings(mainFile.getName().replace(".enso", "") + ".main");
     }
     var context = new PolyglotContext(factory.build());
-
-    if (projectMode) {
-      var result = PackageManager$.MODULE$.Default().loadPackage(file);
-      if (result.isSuccess()) {
-        var s = (scala.util.Success) result;
-        @SuppressWarnings("unchecked")
-        var pkg = (org.enso.pkg.Package<java.io.File>) s.get();
-        var mainModuleName = pkg.moduleNameForFile(pkg.mainFile()).toString();
-        runPackage(context, mainModuleName, file, additionalArgs);
+    try {
+      if (projectMode) {
+        var result = PackageManager$.MODULE$.Default().loadPackage(file);
+        if (result.isSuccess()) {
+          var s = (scala.util.Success) result;
+          @SuppressWarnings("unchecked")
+          var pkg = (org.enso.pkg.Package<java.io.File>) s.get();
+          var mainModuleName = pkg.moduleNameForFile(pkg.mainFile()).toString();
+          runPackage(context, mainModuleName, file, additionalArgs);
+        } else {
+          println(((scala.util.Failure) result).exception().getMessage());
+          throw exitFail();
+        }
       } else {
-        println(((scala.util.Failure) result).exception().getMessage());
-        throw exitFail();
+        runSingleFile(context, file, additionalArgs);
       }
-    } else {
-      runSingleFile(context, file, additionalArgs);
+    } catch (RuntimeException e) {
+      // forces computation of the exception message sooner than context is closed
+      // should work around issues seen at #11127
+      logger.debug("Execution failed with " + e.getMessage());
+      throw e;
+    } finally {
+      context.context().close();
     }
-    context.context().close();
     throw exitSuccess();
   }
 
@@ -1215,7 +1222,7 @@ public class Main {
 
     try {
       return main.call();
-    } catch (IOException ex) {
+    } catch (IOException | RuntimeException ex) {
       throw ex;
     } catch (Exception ex) {
       throw new IOException(ex);

--- a/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/PanicsTest.scala
+++ b/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/PanicsTest.scala
@@ -49,6 +49,29 @@ class PanicsTest extends InterpreterTest {
       consumeOut shouldEqual List("(Error: MyError)")
     }
 
+    "message should be computed soon enough" in {
+      val code =
+        """from Standard.Base import all
+          |
+          |type LM
+          |    V txt
+          |
+          |    to_display_text self =
+          |        res = self.txt + " from exception"
+          |        Panic.throw res
+          |
+          |main = Panic.throw (LM.V "Hi")
+          |""".stripMargin
+
+      try {
+        eval(code)
+      } catch {
+        case ex: InterpreterException =>
+          ex.getMessage() shouldEqual "Hi from exception"
+        case any: Throwable => throw any
+      }
+    }
+
     "catch polyglot errors" in {
       val code =
         """from Standard.Base import all

--- a/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/PanicsTest.scala
+++ b/engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/PanicsTest.scala
@@ -49,7 +49,7 @@ class PanicsTest extends InterpreterTest {
       consumeOut shouldEqual List("(Error: MyError)")
     }
 
-    "message should be computed soon enough" in {
+    "message should be computed before leaving the context" in {
       val code =
         """from Standard.Base import all
           |
@@ -65,9 +65,34 @@ class PanicsTest extends InterpreterTest {
 
       try {
         eval(code)
+        fail("Should raise an InterpreterException");
       } catch {
         case ex: InterpreterException =>
           ex.getMessage() shouldEqual "Hi from exception"
+        case any: Throwable => throw any
+      }
+    }
+
+    "panic causing stack overflow in to_display_text should still generate some error message" in {
+      val code =
+        """from Standard.Base import all
+          |
+          |type LM
+          |    V txt
+          |
+          |    to_display_text self =
+          |        res = LM.V (self.txt + "Ex")
+          |        res.to_display_text
+          |
+          |main = Panic.throw (LM.V "Hi")
+          |""".stripMargin
+
+      try {
+        eval(code)
+        fail("Should raise an InterpreterException");
+      } catch {
+        case ex: InterpreterException =>
+          ex.getMessage() shouldEqual "LM"
         case any: Throwable => throw any
       }
     }

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
@@ -17,6 +17,7 @@ import com.oracle.truffle.api.library.ExportLibrary;
 import com.oracle.truffle.api.library.ExportMessage;
 import com.oracle.truffle.api.nodes.Node;
 import com.oracle.truffle.api.nodes.RootNode;
+import com.oracle.truffle.api.profiles.InlinedBranchProfile;
 import com.oracle.truffle.api.source.SourceSection;
 import org.enso.common.MethodNames;
 import org.enso.interpreter.node.callable.InteropApplicationNode;
@@ -29,6 +30,7 @@ import org.enso.interpreter.runtime.callable.function.FunctionSchema.CallerFrame
 import org.enso.interpreter.runtime.data.EnsoObject;
 import org.enso.interpreter.runtime.data.Type;
 import org.enso.interpreter.runtime.data.vector.ArrayLikeHelpers;
+import org.enso.interpreter.runtime.error.PanicException;
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 import org.enso.interpreter.runtime.state.State;
 import org.enso.interpreter.runtime.type.Types;
@@ -202,7 +204,8 @@ public final class Function implements EnsoObject {
         Function function,
         Object[] arguments,
         @Cached InteropApplicationNode interopApplicationNode,
-        @CachedLibrary("function") InteropLibrary thisLib) {
+        @CachedLibrary("function") InteropLibrary thisLib,
+        @Cached InlinedBranchProfile panicProfile) {
       try {
         return interopApplicationNode.execute(
             function, EnsoContext.get(thisLib).emptyState(), arguments);
@@ -217,6 +220,11 @@ public final class Function implements EnsoObject {
           logger.debug("StackOverflowError detected", err);
         }
         throw err;
+      } catch (PanicException ex) {
+        panicProfile.enter(thisLib);
+        // materialize the exception message
+        ex.getMessage();
+        throw ex;
       }
     }
   }

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java
@@ -32,6 +32,7 @@ import org.enso.interpreter.runtime.data.vector.ArrayLikeHelpers;
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 import org.enso.interpreter.runtime.state.State;
 import org.enso.interpreter.runtime.type.Types;
+import org.slf4j.LoggerFactory;
 
 /** A runtime representation of a function object in Enso. */
 @ExportLibrary(InteropLibrary.class)
@@ -202,8 +203,21 @@ public final class Function implements EnsoObject {
         Object[] arguments,
         @Cached InteropApplicationNode interopApplicationNode,
         @CachedLibrary("function") InteropLibrary thisLib) {
-      return interopApplicationNode.execute(
-          function, EnsoContext.get(thisLib).emptyState(), arguments);
+      try {
+        return interopApplicationNode.execute(
+            function, EnsoContext.get(thisLib).emptyState(), arguments);
+      } catch (StackOverflowError err) {
+        CompilerDirectives.transferToInterpreter();
+        var asserts = false;
+        assert asserts = true;
+        var logger = LoggerFactory.getLogger(Function.class);
+        if (asserts) {
+          logger.error("StackOverflowError detected", err);
+        } else {
+          logger.debug("StackOverflowError detected", err);
+        }
+        throw err;
+      }
     }
   }
 

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -59,6 +59,10 @@ public final class PanicException extends AbstractTruffleException implements En
       throw new IllegalArgumentException("Only interop values are supported: " + payload);
     }
     this.payload = payload;
+    if (CompilerDirectives.inInterpreter()) {
+        getMessage();
+        assert cacheMessage != null;
+    }
   }
 
   /**

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -60,7 +60,7 @@ public final class PanicException extends AbstractTruffleException implements En
       throw new IllegalArgumentException("Only interop values are supported: " + payload);
     }
     this.payload = payload;
-    if (CompilerDirectives.inInterpreter() && location == null) {
+    if (CompilerDirectives.inInterpreter()) {
       cacheMessage = computeMessage();
       assert cacheMessage != null;
     }
@@ -86,13 +86,12 @@ public final class PanicException extends AbstractTruffleException implements En
   @CompilerDirectives.TruffleBoundary
   private String computeMessage() {
     String msg;
-    InteropLibrary library = InteropLibrary.getUncached();
-    Object info = null;
+    var library = InteropLibrary.getUncached();
     try {
-      info = library.getExceptionMessage(this);
+      var info = library.getExceptionMessage(this);
       msg = library.asString(info);
     } catch (AssertionError | UnsupportedMessageException e) {
-      logger().error("Cannot convert " + info + " to string", e);
+      logger().error("Cannot compute message for " + payload, e);
       msg = TypeToDisplayTextNode.getUncached().execute(payload);
     }
     cacheMessage = msg;

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -60,10 +60,6 @@ public final class PanicException extends AbstractTruffleException implements En
       throw new IllegalArgumentException("Only interop values are supported: " + payload);
     }
     this.payload = payload;
-    if (CompilerDirectives.inInterpreter()) {
-      cacheMessage = computeMessage();
-      assert cacheMessage != null;
-    }
   }
 
   /**

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -13,7 +13,6 @@ import com.oracle.truffle.api.library.ExportLibrary;
 import com.oracle.truffle.api.library.ExportMessage;
 import com.oracle.truffle.api.nodes.Node;
 import com.oracle.truffle.api.source.SourceSection;
-import java.util.logging.Level;
 import org.enso.interpreter.node.BaseNode.TailStatus;
 import org.enso.interpreter.node.callable.IndirectInvokeMethodNode;
 import org.enso.interpreter.node.callable.InvokeCallableNode.ArgumentsExecutionMode;
@@ -27,6 +26,8 @@ import org.enso.interpreter.runtime.data.Type;
 import org.enso.interpreter.runtime.data.text.Text;
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 import org.enso.interpreter.runtime.state.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** An exception type for user thrown panic exceptions. */
 @ExportLibrary(value = InteropLibrary.class, delegateTo = "payload")
@@ -59,8 +60,8 @@ public final class PanicException extends AbstractTruffleException implements En
       throw new IllegalArgumentException("Only interop values are supported: " + payload);
     }
     this.payload = payload;
-    if (CompilerDirectives.inInterpreter()) {
-      getMessage();
+    if (CompilerDirectives.inInterpreter() && location == null) {
+      cacheMessage = computeMessage();
       assert cacheMessage != null;
     }
   }
@@ -91,15 +92,8 @@ public final class PanicException extends AbstractTruffleException implements En
       info = library.getExceptionMessage(this);
       msg = library.asString(info);
     } catch (AssertionError | UnsupportedMessageException e) {
-      try {
-        var ctx = EnsoContext.get(null);
-        ctx.getLogger().log(Level.WARNING, "Cannot convert " + info + " to string", e);
-        msg = TypeToDisplayTextNode.getUncached().execute(payload);
-      } catch (AssertionError assertionError) {
-        throw new AssertionError(
-            "Failed to log failed conversion of " + info + " to string and payload " + payload,
-            assertionError);
-      }
+      logger().error("Cannot convert " + info + " to string", e);
+      msg = TypeToDisplayTextNode.getUncached().execute(payload);
     }
     cacheMessage = msg;
     return msg;
@@ -169,7 +163,7 @@ public final class PanicException extends AbstractTruffleException implements En
       return Text.create(strings.asString(text));
     } catch (UnsupportedMessageException e) {
       CompilerDirectives.transferToInterpreter();
-      ctx.getLogger().log(Level.WARNING, "Cannot convert " + text + " to string", e);
+      logger().error("Cannot convert " + text + " to string", e);
       return Text.create(typeToDisplayTextNode.execute(payload));
     }
   }
@@ -221,5 +215,9 @@ public final class PanicException extends AbstractTruffleException implements En
       throw UnsupportedMessageException.create();
     }
     return getLocation().getEncapsulatingSourceSection();
+  }
+
+  private static Logger logger() {
+    return LoggerFactory.getLogger(PanicException.class);
   }
 }

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -60,8 +60,8 @@ public final class PanicException extends AbstractTruffleException implements En
     }
     this.payload = payload;
     if (CompilerDirectives.inInterpreter()) {
-        getMessage();
-        assert cacheMessage != null;
+      getMessage();
+      assert cacheMessage != null;
     }
   }
 

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicException.java
@@ -86,8 +86,8 @@ public final class PanicException extends AbstractTruffleException implements En
     try {
       var info = library.getExceptionMessage(this);
       msg = library.asString(info);
-    } catch (AssertionError | UnsupportedMessageException e) {
-      logger().error("Cannot compute message for " + payload, e);
+    } catch (StackOverflowError | AssertionError | UnsupportedMessageException e) {
+      logger().atError().log("Cannot compute message for " + payload, e);
       msg = TypeToDisplayTextNode.getUncached().execute(payload);
     }
     cacheMessage = msg;

